### PR TITLE
Add MobileViTv2

### PR DESCRIPTION
**Pull request by shehanmunasinghe**
# What does this PR do?

Adds the MobileViTv2 model into transformers library (PS: Work in Progress)

Fixes # ([issue](https://github.com/huggingface/transformers/issues/22570))

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/22570 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

File: docs/source/en/serialization.mdx [hunk 1]
```diff
@@ -99,6 +99,7 @@ Ready-made configurations include the following architectures:
 - MobileNetV1
 - MobileNetV2
 - MobileViT
+- MobileViTv2v2
```
Comment (amyeroberts) [inline docs/source/en/serialization.mdx:102]: Typo?

```suggestion
- MobileViTV2
```

File: src/transformers/models/mobilevitv2/image_processing_mobilevitv2.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        resample: PILImageResampling = None,
+        do_rescale: bool = None,
+        rescale_factor: float = None,
+        do_center_crop: bool = None,
+        crop_size: Dict[str, int] = None,
+        do_flip_channel_order: bool = None,
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        data_format: ChannelDimension = ChannelDimension.FIRST,
+        **kwargs,
+    ) -> PIL.Image.Image:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/image_processing_mobilevitv2.py:234]:
```suggestion
    ) -> BatchFeature:
```

File: tests/models/mobilevitv2/test_modeling_mobilevitv2.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    Here we also overwrite some of the tests of test_modeling_common.py, as MobileViTv2 does not use input_ids, inputs_embeds,
+    attention_mask and seq_length.
+    """
+
+    all_model_classes = (
+        (MobileViTv2Model, MobileViTv2ForImageClassification, MobileViTv2ForSemanticSegmentation)
+        if is_torch_available()
+        else ()
+    )
+
```
Comment (amyeroberts) [inline tests/models/mobilevitv2/test_modeling_mobilevitv2.py:188]: I think you'll need to add `pipeline_model_mapping` here e.g. like [here](https://github.com/huggingface/transformers/blob/d95045717e1a5bd8ce71223b5b8920e27687dee4/tests/models/mobilevit/test_modeling_mobilevit.py#L188)

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! 

Looking at the modeling file -all of the layers are copied from the original MobileVit. Is this correct? If this is the case, then we can do something similar to models like [Matcha](https://github.com/huggingface/transformers/blob/main/docs/source/en/model_doc/matcha.mdx), where we add the model in the README and checkpoints are loaded in using the Pix2Struct architecture, but no new modeling file is added.

**Comment by shehanmunasinghe** [comment]
@amyeroberts please note that this is still a work in progress. I'll let you know when it is ready for your review.

**Comment by shehanmunasinghe** [comment]
@amyeroberts , this is now ready for your review.

**Comment by shehanmunasinghe** [comment]
@amyeroberts , any updates?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 4]
```diff
@@ -0,0 +1,1200 @@
...
+    if min_value is None:
+        min_value = divisor
+    new_value = max(min_value, int(value + divisor / 2) // divisor * divisor)
+    # Make sure that round down does not go down by more than 10%.
+    if new_value < 0.9 * value:
+        new_value += divisor
+    return int(new_value)
+
+
+def bound_fn(min_val: Union[float, int], max_val: Union[float, int], value: Union[float, int]) -> Union[float, int]:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:81]: * `clip` would be a better name here. `bound_fn` sounds like the function has been bound to something
* Let's put `value` as the first value to keep with similar functions torch and numpy's clip 
* For typing - if both float and ints are accepted input types,[ just `float` can be used](https://peps.python.org/pep-0484/#the-numeric-tower)

```suggestion
def clip(value: float, min_val: float = float("-inf"), max_val: float = float("inf")) -> float:
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 5]
```diff
@@ -0,0 +1,1200 @@
...
+        on resource-constrained devices) that may be required to convert the unfolded tensor from channel-first to
+        channel-last format in case of a linear layer.
+    """
+
+    def __init__(
+        self,
+        config: MobileViTv2Config,
+        embed_dim: int,
+        attn_dropout: Optional[float] = 0.0,
+        bias: Optional[bool] = True,
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:286]: Why have this arg if it's always True?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 6]
```diff
@@ -0,0 +1,1200 @@
...
+        point-wise convolution (instead of a linear layer). This avoids a transpose operation (which may be expensive
+        on resource-constrained devices) that may be required to convert the unfolded tensor from channel-first to
+        channel-last format in case of a linear layer.
+    """
+
+    def __init__(
+        self,
+        config: MobileViTv2Config,
+        embed_dim: int,
+        attn_dropout: Optional[float] = 0.0,
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:285]: This should be take from the passed in the config 
```suggestion
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 7]
```diff
@@ -0,0 +1,1200 @@
...
+            self.layer.append(layer)
+            in_channels = out_channels
+
+    def forward(self, features: torch.Tensor) -> torch.Tensor:
+        for layer_module in self.layer:
+            features = layer_module(features)
+        return features
+
+
+class MobileViTv2LinearSelfAttention(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:255]: It's confusing to have a layer called `SelfAttention` which performs also cross attention - it should either be just `Attention` or two separate layers. Look to other models in the library to see how self/cross attention layers logic is normally handled

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 8]
```diff
@@ -0,0 +1,1200 @@
...
+        # 3: layer_3
+        layer_3 = MobileViTv2Layer(
+            config,
+            in_channels=layer_2_dim,
+            out_channels=layer_3_dim,
+            attn_unit_dim=int(make_divisible(128 * config.width_multiplier, divisor=8)),
+            stride=2,
+            ffn_multiplier=ffn_multiplier,
+            n_attn_blocks=2,
+        )
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:684]: For all of the `MobileViTv2Layer` layer here and all below: 
* `attn_unit_dim` shouldn't be calculated form hard coded values here e.g. 128 and divisor etc. These values should be part of the configuration e.g. `make_divisible(config.x * config.width_multiplier, divisor=config.y),`
* `stride` argument can be remove - it's 2 for everything
* `n_attn_blocks` should be in the config 
* `patch_size` is taken from the config - arg should be remove and config used instead 
* `ffn_multiplier` can be removed as an arg - its 2 everywhere (and should come from the config)

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 9]
```diff
@@ -0,0 +1,1200 @@
...
+        dilate_layer_4 = dilate_layer_5 = False
+        if config.output_stride == 8:
+            dilate_layer_4 = True
+            dilate_layer_5 = True
+        elif config.output_stride == 16:
+            dilate_layer_5 = True
+
+        dilation = 1
+
+        layer_0_dim = int(
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:643]: You don't need to wrap all of the `make_divisible` call with `int(...)` - the function already handles that

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 10]
```diff
@@ -0,0 +1,1200 @@
...
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tabby, tabby cat"
+
+
+MOBILEVITV2_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "apple/mobilevitv2-1.0-imagenet1k-256"
+    # See all MobileViTv2 models at https://huggingface.co/models?filter=mobilevitv2
+]
+
+
+def make_divisible(value: int, divisor: int = 8, min_value: Optional[int] = None) -> int:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:66]: This should have a copied from statement

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 11]
```diff
@@ -0,0 +1,1200 @@
...
+            out_channels=in_channels,
+            kernel_size=1,
+            use_normalization=True,
+            use_activation=False,
+        )
+
+        self.patch_h = patch_size
+        self.patch_w = patch_size
+
+    def unfolding_pytorch(self, feature_map: torch.Tensor) -> Tuple[torch.Tensor, Tuple[int, int]]:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:570]: Why specify `pytorch` in the name here - the whole model is in pytorch

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 12]
```diff
@@ -0,0 +1,1200 @@
...
+        batch_size, in_channels, img_h, img_w = feature_map.shape
+
+        # [B, C, H, W] --> [B, C, P, N]
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:573]: clearer var names

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 13]
```diff
@@ -0,0 +1,1200 @@
...
+        patches = nn.functional.unfold(
+            feature_map,
+            kernel_size=(self.patch_h, self.patch_w),
+            stride=(self.patch_h, self.patch_w),
+        )
+        patches = patches.reshape(batch_size, in_channels, self.patch_h * self.patch_w, -1)
+
+        return patches, (img_h, img_w)
+
+    def folding_pytorch(self, patches: torch.Tensor, output_size: Tuple[int, int]) -> torch.Tensor:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:583]: Same here re method name

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 14]
```diff
@@ -0,0 +1,1200 @@
...
+        batch_size, in_dim, patch_size, n_patches = patches.shape
+
+        # [B, C, P, N]
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:586]: same here re clearer var names

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 15]
```diff
@@ -0,0 +1,1200 @@
...
+            features = self.downsampling_layer(features)
+
+        # local representation
+        features = self.conv_kxk(features)
+        features = self.conv_1x1(features)
+
+        # convert feature map to patches
+        patches, output_size = self.unfolding_pytorch(
+            features,
+        )
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:610]:
```suggestion
        patches, output_size = self.unfolding_pytorch(features)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! 

Apologies for the delay in review. I've done a first pass on the files, generally in a good shape and very clean code 🤗  There's a bit of reworking needing in the attention layer so will need likely need more than one final review to make sure all in line with the library patterns. 

General comments: 
* Model name: MobileVitV2 to be consistent with other models + camelcasing
* Individual layers shouldn't have *args and **kwargs 
* __repr__ implementations aren't needed for model classes, particularly private ones
* Explicit var names in code and in comments e.g. N -> batch_size
* Typically don't use asserts in the main modeling code - raise exceptions instead or remove
* Feature extractors are deprecated for vision models
* Use of config values and matching with the model implementation - specifically the activations don't seem to match
* Layers should, as much as possible, not have kwargs for their init and just use the config to instantiate themselves. There's a few values in the modeling file taken form the config and then passed along to all the layers e.g.

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 16]
```diff
@@ -0,0 +1,1200 @@
...
+            raise ValueError("Expected 3 values for atrous_rates")
+
+        self.convs = nn.ModuleList()
+
+        in_projection = MobileViTv2ConvLayer(
+            config,
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=1,
+            use_activation="relu",
```
[the remark this answers is not in the document]
Comment (shehanmunasinghe) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:1039]: Activation from the config file is for the base model. This line of code is for the segmentation head, which uses a different activation. Here I'm following the MobileViT(v1) implementation.

File: README_zh-hans.md [hunk 17]
```diff
@@ -339,7 +339,7 @@ conda install -c huggingface transformers
...
 1. **[MarkupLM](https://huggingface.co/docs/transformers/model_doc/markuplm)** (来自 Microsoft Research Asia) 伴随论文 [MarkupLM: Pre-training of Text and Markup Language for Visually-rich Document Understanding](https://arxiv.org/abs/2110.08518) 由 Junlong Li, Yiheng Xu, Lei Cui, Furu Wei 发布。
 1. **[Mask2Former](https://huggingface.co/docs/transformers/model_doc/mask2former)** (来自 FAIR and UIUC) 伴随论文 [Masked-attention Mask Transformer for Universal Image Segmentation](https://arxiv.org/abs/2112.01527) 由 Bowen Cheng, Ishan Misra, Alexander G. Schwing, Alexander Kirillov, Rohit Girdhar 发布。
-1. **[MaskFormer](https://huggingface.co/docs/transformers/model_doc/maskformer)** (from Meta and UIUC) released with the paper [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov
+1. **[MaskFormer](https://huggingface.co/docs/transformers/model_doc/maskformer)** (from Meta and UIUC) released with the paper [Per-Pixel Classification is Not All You Need for Semantic Segmentation](https://arxiv.org/abs/2107.06278) by Bowen Cheng, Alexander G. Schwing, Alexander Kirillov
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply README_zh-hans.md:342]: This hasn't been resolved

File: src/transformers/models/auto/image_processing_auto.py [hunk 18]
```diff
@@ -77,6 +77,8 @@
         ("mobilenet_v2", "MobileNetV2ImageProcessor"),
         ("mobilevit", "MobileViTImageProcessor"),
         ("mobilevit", "MobileViTImageProcessor"),
+        ("mobilevitv2", "MobileViTImageProcessor"),
```
Comment (amyeroberts) [inline src/transformers/models/auto/image_processing_auto.py:80]:
```suggestion
```

File: src/transformers/models/mobilevitv2/__init__.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+try:
+    if not is_vision_available():
+        raise OptionalDependencyNotAvailable()
+except OptionalDependencyNotAvailable:
+    pass
+else:
+    pass
+
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/__init__.py:39]:
```suggestion
```

File: src/transformers/models/mobilevitv2/__init__.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    )
+
+    try:
+        if not is_vision_available():
+            raise OptionalDependencyNotAvailable()
+    except OptionalDependencyNotAvailable:
+        pass
+    else:
+        pass
+
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/__init__.py:68]:
```suggestion

```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 21]
```diff
@@ -0,0 +1,1073 @@
...
+        - Input: :math:`(batch_size, input_channels, num_pixels_in_patch, num_patches)`
+        - Output: same as the input
+
+    """
+
+    def __init__(
+        self,
+        config: MobileViTV2Config,
+        embed_dim: int,
+    ) -> None:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:244]: nit: can be put on one line 

```suggestion
    def __init__(self, config: MobileViTV2Config, embed_dim: int) -> None:
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 22]
```diff
@@ -0,0 +1,1073 @@
...
+
+    Args:
+        config: MobileVitv2Config
+        embed_dim (int):
+            :math:`input_channels` from an expected input of size :math:`(batch_size, input_channels, height, width)`
+
+    Shape:
+        - Input: :math:`(batch_size, input_channels, num_pixels_in_patch, num_patches)`
+        - Output: same as the input
+
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:237]:
```suggestion
        config (`MobileVitv2Config`):
             Model configuration object
        embed_dim (`int`):
            `input_channels` from an expected input of size :math:`(batch_size, input_channels, height, width)`
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 23]
```diff
@@ -0,0 +1,1073 @@
...
+        context_vector = torch.sum(context_vector, dim=-1, keepdim=True)
+
+        # combine context vector with values
+        # [batch_size, embed_dim, num_pixels_in_patch, num_patches] * [batch_size, embed_dim, num_pixels_in_patch, 1] --> [batch_size, embed_dim, num_pixels_in_patch, num_patches]
+        out = torch.nn.functional.relu(value) * context_vector.expand_as(value)
+        out = self.out_proj(out)
+        return out
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        return self._forward_self_attn(x, *args, **kwargs)
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:295]: - All of the `_forward_self_attn` logic should go here and `_forward_self_attn` removed (it's superfluous)
- Forward arguments shouldn't take arbirary *args and **kwargs
- Rename `x` to `hidden_states` to be consistent with other `forward` methods

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 24]
```diff
@@ -0,0 +1,1073 @@
...
+    def __init__(
+        self,
+        config: MobileViTV2Config,
+        embed_dim: int,
+        ffn_latent_dim: int,
+        dropout: float = 0.0,
+    ) -> None:
+        super().__init__()
+        # pre_norm_attn
+        # self.layernorm_before = MobileViTV2LayerNorm2D(embed_dim, eps=config.layer_norm_eps)
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:349]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 25]
```diff
@@ -0,0 +1,1073 @@
...
+        self.layernorm_before = nn.GroupNorm(num_groups=1, num_channels=embed_dim, eps=config.layer_norm_eps)
+        self.attention = MobileViTV2LinearSelfAttention(config, embed_dim)
+        self.dropout1 = nn.Dropout(p=dropout)
+        # pre_norm_ffn
+        # self.layernorm_after = MobileViTV2LayerNorm2D(embed_dim, eps=config.layer_norm_eps)
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:354]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 26]
```diff
@@ -0,0 +1,1073 @@
...
+        self,
+        config: MobileViTV2Config,
+        n_layers: int,
+        d_model: int,
+    ) -> None:
+        super().__init__()
+
+        ffn_multiplier = config.ffn_multiplier
+
+        # if isinstance(ffn_multiplier, Sequence) and len(ffn_multiplier) == 2:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:381]: To remove

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 27]
```diff
@@ -0,0 +1,1073 @@
...
+        return layer_output
+
+
+class MobileViTV2Transformer(nn.Module):
+    def __init__(
+        self,
+        config: MobileViTV2Config,
+        n_layers: int,
+        d_model: int,
+    ) -> None:
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:376]: nit: can be put on one line

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 28]
```diff
@@ -0,0 +1,1073 @@
...
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        for layer_module in self.layer:
+            hidden_states = layer_module(hidden_states)
+        return hidden_states
+
+
+class MobileViTV2Layer(nn.Module):
+    """
+    MobileViTV2 block: https://arxiv.org/abs/2206.02680
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:410]: Perhaps we should rename this class to `MobileViTV2Block` - I'll let you decide what you think is best

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 29]
```diff
@@ -0,0 +1,1073 @@
...
+    """
+
+    def __init__(
+        self,
+        config: MobileViTV2Config,
+        in_channels: int,
+        out_channels: int,
+        attn_unit_dim: int,
+        stride: Optional[int] = 2,
+        n_attn_blocks: Optional[int] = 2,
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:420]: Can these ever have the value `None` ?

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 30]
```diff
@@ -0,0 +1,1073 @@
...
+    def unfolding(self, feature_map: torch.Tensor) -> Tuple[torch.Tensor, Tuple[int, int]]:
+        batch_size, in_channels, img_h, img_w = feature_map.shape
+        patches = nn.functional.unfold(
+            feature_map,
+            kernel_size=(self.patch_height, self.patch_width),
+            stride=(self.patch_height, self.patch_width),
+        )
+        patches = patches.reshape(batch_size, in_channels, self.patch_height * self.patch_width, -1)
+
+        return patches, (img_h, img_w)
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:483]: more explicit var names

```suggestion
        batch_size, in_channels, img_height, img_width = feature_map.shape
        patches = nn.functional.unfold(
            feature_map,
            kernel_size=(self.patch_height, self.patch_width),
            stride=(self.patch_height, self.patch_width),
        )
        patches = patches.reshape(batch_size, in_channels, self.patch_height * self.patch_width, -1)

        return patches, (img_height, img_width)
```

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py [hunk 31]
```diff
@@ -0,0 +1,1073 @@
...
+
+        # convert feature map to patches
+        patches, output_size = self.unfolding(features)
+
+        # learn global representations
+        patches = self.transformer(patches)
+        patches = self.layernorm(patches)
+
+        # convert patches back to feature maps
+        # [B x Patch x Patches x C] --> [B x C x Patches x Patch]
```
Comment (amyeroberts) [inline src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:515]: more explicit var names here in place of single letters i.e. B -> batch_size

**Review by amyeroberts — COMMENTED** [review]
Looking good! 

Just a small few changes and then it's good to merge. Overall clean and easy to read - just make sure to remove all old comments and rework any logic where it makes sense e.g. removing `_forward_self_attn` now the method is redundant.

File: src/transformers/models/mobilevitv2/modeling_mobilevitv2.py (hunk 28 above)
Comment (shehanmunasinghe) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py:410]: I think using `MobileViTV2Layer` is more consistent with the original MobileViT (v1) implementation.

Comment (amyeroberts) [reply src/transformers/models/mobilevitv2/modeling_mobilevitv2.py]: Sounds good!

**Comment by shehanmunasinghe** [comment]
@amyeroberts, thanks for your feedback and I have now applied the suggested changes.

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this model!

**Pull request merged**